model/dns_zone/dns_zone.rb

```diff
@@ -31,5 +37,50 @@ def delete_record(record_name:, type: nil, data: nil)
     records.each do |r|
       r.destroy
     end
+
+    command = "zone-unset #{name} #{record_name} #{type} #{data}"
+    run_on_all_name_servers(command, "no such node in zone found")
+  end
+
+  def run_on_all_name_servers(command, ignore)
+    lock_name_servers do
+      name_servers.flat_map(&:vms).each do |vm|
+        # knotc also supports passing multiple commands via stdin, but we
```

Comment:
I don't 100% follow the implications of this comment. Isn't he point of the begin/commit + stdin thing is to have simpler atomic handling?  I could understand, perhaps, if I saw interesting branching going on for each knotc command, but I don't.

I can kind of understand the case of the first abort, this effectively has a branch, where abort = success, abort failed because it doesn't exist = success, everything else raises.

although, wouldn't a dangling transaction be unexpected, and noteworthy?

Reply:
knotc's transactions are bit weird and not quite sophisticated as PostgreSQL's. Here are few interesting properties;

- Sending commands separately or at a single time via stdin has same effect. Only difference is error handling. If you send it via stdin, it always exit with 0, so you need to scan the stdout for errors. If you send each command separately, they can individually exit with an error code and it is easier to do error handling.
- Transactions are independent from connections. You can do begin in one shell and commit in another.
- As corollary of above, there can be only one active transaction (per zone)
- Transactions can be committed partially. Let's say you run multiple commands and some of them succeeded. If you commit, the succeeded commands would take effect.

Dangling transaction wouldn't be very rare. Any time an error is raised, we exit and the next iteration would get a dangling transaction.If an error happens, it is user's responsibility to abort. An alternative solutions would be catching all exceptions and running an abort then, but what if abort itself fail. In any case, we should run an abort before doing anything. We got the lock, so if there is anything left dangling it should be safe to abort.

Reply:
I switched to stdin, which would require some error handling based on stdout. I'll add that as well.